### PR TITLE
AzDO Migration - Move off self-hosted pools, build script improvements, naming cleanups 

### PR DESCRIPTION
**Pull request by jimson-msft**
This is part 1 of the migration work to use 1ES Hosted agents:
- Migrate to using microsoft-hosted agents and 1es hosted agents in all pipeline yaml files
- Converge to using a single bootstrap script
- Various cleanup around paths and naming conventions

TODO: 
Native builds are using a 1ES hosted image of a blank Ubuntu 18.04 image, we can use ImageFactory artifacts to preinstall our dependencies to reduce COGS
Docker builds are using Microsoft-hosted agents which contain docker preinstalled, this is because trying to install and use docker on a blank ubuntu 18.04 image leads to permissions issues with the az pipelines agent calling the docker cli

See: https://deliveryoptimization.visualstudio.com/client/_build
All runs succesful

### CODE REVIEW

File: build/bootstrap.sh [hunk 1]
```diff
@@ -0,0 +1,220 @@
+#! /bin/bash
```
Comment (jimson-msft) [inline build/bootstrap.sh:1]: Part of the motivation behind this script change is to be able to invoke it as an image factory artifact

File: azure-pipelines/build/templates/dosdkcpp-docker-steps.yml [hunk 2]
```diff
@@ -2,18 +2,47 @@
 # Consume this steps template in one or more jobs by passing in parameter values.
 
 parameters:
-- name: targetOsArch    # example: debian10-arm32
+- name: targetOS        # example: ubuntu18.04
```
Comment (jimson-msft) [inline azure-pipelines/build/templates/dosdkcpp-docker-steps.yml:5]: >ubuntu18.04 [](http://example.com/codeflow?start=35&length=11)

nit: ubuntu1804, same with line 7

File: azure-pipelines/publishing/github-release.yml [hunk 3]
```diff
@@ -6,7 +6,7 @@ trigger:
 pr:
 - none
 
-pool: dotestlab
+pool: 1es_hosted_pool_ubuntu_1804
```
Comment (jimson-msft) [inline azure-pipelines/publishing/github-release.yml:9]: This should actually be using the msft hosted agent, until 1ES machine can support docker

File: build/bootstrap.sh [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+
+###
+# This script handles provisioning of the Delivery Optimization client components on supported platforms
+###
+
+# bootstrap scripts will exit immediately if a command exits with a non-zero status
```
Comment (jimson-msft) [inline build/bootstrap.sh:7]: can remove this comment

File: build/docker/debian10/arm32/Dockerfile [hunk 5]
```diff
@@ -17,25 +17,25 @@ SHELL [ "/bin/bash", "-c"]
...
-RUN chmod +x bootstrap-debian-10.sh
-RUN ./bootstrap-debian-10.sh
+RUN chmod +x bootstrap.sh
+RUN ./bootstrap.sh --platform debian10 --install build
 
 VOLUME /code
 WORKDIR /code
 
-ENTRYPOINT [ "/bin/bash", "-c" ]
+ENTRYPOINT [ "/bin/bash", "-c"]
```
Comment (shishirb-MSFT) [inline build/docker/debian10/arm32/Dockerfile:38]: >] [](http://example.com/codeflow?start=30&length=1)

nit: revert the change (spacing)

File: build/bootstrap.sh (hunk 4 above)
Comment (shishirb-MSFT) [reply build/bootstrap.sh:7]: Maybe keep it and shorten it to 'Treat all command failures as fatal'

File: build/bootstrap.sh [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+set -e
+
+# Defaults 
+INSTALL=all
+
+usage() {
+    cat <<EOM
+$(basename $0) - Script to setup development environments for Delivery Optimization
+Usage: $(basename $0) --platform <platform to install for> --install <install command>
+    --platform          # Platform to provision, supported platforms: Ubuntu1804, Ubuntu 2004, Debian9, Debian 10. Default is Ubuntu1804
```
Comment (jimson-msft) [inline build/bootstrap.sh:17]: >  [](http://example.com/codeflow?start=110&length=1)

nit: space

File: build/bootstrap.sh [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+    --install           # Which command to run, supported commands: builddependencies, developertools, containertools, qemu, all. Default is All
+EOM
+    exit 1
+}
+
+function parseArgs() {
```
Comment (jimson-msft) [inline build/bootstrap.sh:23]: >function parseArgs() [](http://example.com/codeflow?start=0&length=20)

not the most familiar with shell scripting, but I think the () can be removed here, usage() can also be a function to stay consistent

File: build/bootstrap.sh [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+            exit 0
+            ;;
+        --platform | -p)
+            PLATFORM="${2,,}"
+            if [[ "$PLATFORM" == "debian9" || "$PLATFORM" == "debian10" || "$PLATFORM" == "ubuntu1804" || "$PLATFORM" == "ubuntu2004" ]];
+            then
+                echo -e "[INFO] Platform set to: ${PLATFORM}"
+            else 
+                echo -e "[ERROR] Unsupported platform: ${PLATFORM}"
+                exit
```
Comment (shishirb-MSFT) [inline build/bootstrap.sh:39]: >exit [](http://example.com/codeflow?start=16&length=4)

Exit with an error?

File: build/bootstrap.sh (hunk 6 above)
Comment (shishirb-MSFT) [inline build/bootstrap.sh:17]: ># Platform to provision, supported platforms: Ubuntu1804, Ubuntu 2004, Debian9, Debian 10. Default is Ubuntu1804 [](http://example.com/codeflow?start=24&length=112)

Do these comments get printed to the terminal? If not, consider changing them to normal strings that get printed.

Comment (jimson-msft) [reply build/bootstrap.sh:17]: Yes they get printed

File: build/bootstrap.sh [hunk 9]
```diff
@@ -0,0 +1,220 @@
+#! /bin/bash
+
+###
+# This script handles provisioning of the Delivery Optimization client components on supported platforms
```
Comment (jimson-msft) [inline build/bootstrap.sh:4]: modify this to read "...handles provisiong of build environment for the Delivery Optimization..."

File: build/bootstrap.sh [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+}
+
+function installBuildDependencies
+{
+    if [[ -v PLATFORM ]]
+    then
+        echo "[INFO] Platform check succesful"
+    else
+        echo "[WARNING] No platform supplied, using default: Ubuntu1804"
+        PLATFORM = "ubuntu1804"
```
Comment (shishirb-MSFT) [inline build/bootstrap.sh:64]: Consider not having a default. It is important enough to explicitly specify.

File: build/bootstrap.sh [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+    apt-get install -y libboost-random-dev libboost-regex-dev
+    apt-get install -y libproxy-dev libssl-dev uuid-dev
+
+    # Install cpprest dependencies
+    # libssl-dev also required but installed above because plugin uses libssl-dev directly
+    apt-get install -y zlib1g-dev
+
+    # Most target platforms do not natively have a version of cpprest that supports url-redirection
+    # Build and install v2.10.16 as it's the earliest version which supports url-redirection
+    mkdir /tmp/cpprestsdk
```
Comment (shishirb-MSFT) [inline build/bootstrap.sh:108]: >mkdir /tmp/cpprestsdk [](http://example.com/codeflow?start=4&length=21)

Script will fail if directory already exists. I've been bitten by this. Consider deleting the directory if exists and then continue with creating the dir.

Same for /tmp/gtest and other custom dirs we use.

File: build/bootstrap.sh [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+
+# Defaults 
+INSTALL=all
+
+usage() {
+    cat <<EOM
+$(basename $0) - Script to setup development environments for Delivery Optimization
+Usage: $(basename $0) --platform <platform to install for> --install <install command>
+    --platform          # Platform to provision, supported platforms: Ubuntu1804, Ubuntu 2004, Debian9, Debian 10. Default is Ubuntu1804
+    --install           # Which command to run, supported commands: builddependencies, developertools, containertools, qemu, all. Default is All
```
Comment (shishirb-MSFT) [inline build/bootstrap.sh:18]: >All [](http://example.com/codeflow?start=141&length=3)

all (lowercase)

File: build/bootstrap.sh (hunk 6 above)
Comment (shishirb-MSFT) [reply build/bootstrap.sh:17]: Ok, then change the casing on the platform strings (all lowercase).

File: azure-pipelines/build/templates/dopapt-docker-steps.yml [hunk 13]
```diff
@@ -2,19 +2,52 @@
...
-    script: 'sudo docker run --rm --entrypoint=/bin/bash -v $(Build.SourcesDirectory):/code -v /tmp/build-deliveryoptimization-plugin-apt-${{parameters.targetOsArch}}:/tmp/build-deliveryoptimization-plugin-apt ${{parameters.dockerImageName}} "/code/build/docker/docker-build-plugin.sh" "/code" "${{parameters.config}}"'
+    script: 'chmod +x bootstrap.sh'
+    workingDirectory: '$(Build.SourcesDirectory)/build'
+  displayName: 'chmod bootstrap script'
+
+- task: CmdLine@2
+  inputs:
+    script: 'sudo ./bootstrap.sh --platform ${{parameters.targetOS}} --install containertools'
+    workingDirectory: '$(Build.SourcesDirectory)/build'
+  displayName: 'Install Docker'
```
Comment (shishirb-MSFT) [inline azure-pipelines/build/templates/dopapt-docker-steps.yml:26]: >displayName: 'Install Docker' [](http://example.com/codeflow?start=2&length=29)

Why does only the plugin arm build require docker to be installed?

File: azure-pipelines/build/templates/dosdkcpp-native-steps.yml [hunk 14]
```diff
@@ -49,12 +64,10 @@ steps:
     workingDirectory: '/tmp/build-deliveryoptimization-sdk/linux-${{parameters.config}}'
   displayName: 'Install libdeliveryoptimization Debian Package'
 
-- task: PythonScript@0
-  condition: eq('${{parameters.skipTests}}', false)
```
Comment (shishirb-MSFT) [inline azure-pipelines/build/templates/dosdkcpp-native-steps.yml:53]: Why remove this condition? No need to build tests in release mode.

File: build/bootstrap.sh [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+    installDeveloperTools
+    installContainerTools
+    installQemu
+}
+
+main()
+{
+    parseArgs "$@"
+    
+
```
Comment (jimson-msft) [inline build/bootstrap.sh:193]: nit: extra space

File: azure-pipelines/build/templates/dopapt-docker-steps.yml (hunk 13 above)
Comment (jimson-msft) [reply azure-pipelines/build/templates/dopapt-docker-steps.yml:26]: Oops, forgot to remove

File: azure-pipelines/build/templates/dosdkcpp-native-steps.yml (hunk 14 above)
Comment (jimson-msft) [reply azure-pipelines/build/templates/dosdkcpp-native-steps.yml:53]: oops, good catch

**Comment by jimson-msft** [comment]
> 

Oops, need to undo

---
Refers to: azure-pipelines/build/templates/doclient-lite-docker-steps.yml:26 in 7933105. [](commit_id = 793310581e05631f9414fab3f15bc948dde8d51b, deletion_comment = True)

File: azure-pipelines/build/templates/doclient-lite-docker-steps.yml [hunk 16]
```diff
@@ -2,18 +2,46 @@
...
+- name: targetOS        # example: ubuntu1804
   type: string
-- name: dockerImageName # example: jimsonmsft/debian10-arm32:latest
+- name: targetOsArch    # example: ubuntu1804-x64
   type: string
 - name: config          # debug/release/minsizerel/relwithdebuginfo
   type: string
 
 steps:
+- task: Docker@2
```
Comment (shishirb-MSFT) [inline azure-pipelines/build/templates/doclient-lite-docker-steps.yml:13]: I think all the steps from login to install & register qemu can be moved to a new template and referenced from do-docker-jobs.yml.

Not urgent, feel free to open a task.

**Review by shishirb-MSFT — APPROVED** [review]
:shipit:

File: azure-pipelines/build/templates/doclient-lite-docker-steps.yml (hunk 16 above)
Comment (jimson-msft) [reply azure-pipelines/build/templates/doclient-lite-docker-steps.yml:13]: Well, the registering of QEMU will happen during image provisioning (once this script is uploaded as an artifact to Image Factory), which leaves just the login and pull steps, which could be templatized further, but at that point I'm not sure if it's worth it for two tasks

Comment (shishirb-MSFT) [reply azure-pipelines/build/templates/doclient-lite-docker-steps.yml:13]: Agreed, may not be worth it for just a couple of steps.
